contracts/zero-ex/contracts/src/features/MetaTransactionsFeatureV2.sol

```diff
@@ -0,0 +1,620 @@
+        state.signature = signature;
+
+        returnResult = _executeMetaTransactionPrivate(state);
+    }
+
+    /// @dev Execute multiple meta-transactions.
+    /// @param mtxs The meta-transactions.
+    /// @param signatures The signature by each respective `mtx.signer`.
+    /// @return returnResults The ABI-encoded results of the underlying calls.
+    function batchExecuteMetaTransactionsV2(
+        MetaTransactionDataV2[] memory mtxs,
+        LibSignature.Signature[] memory signatures
+    )
+        public
+        payable
+        override
+        nonReentrant(REENTRANCY_MTX)
+        doesNotReduceEthBalance
+        refundsAttachedEth
+        returns (bytes[] memory returnResults)
+    {
+        if (mtxs.length != signatures.length) {
+            LibMetaTransactionsRichErrors
+                .InvalidMetaTransactionsArrayLengthsError(mtxs.length, signatures.length)
+                .rrevert();
+        }
+        returnResults = new bytes[](mtxs.length);
+        for (uint256 i = 0; i < mtxs.length; ++i) {
+            ExecuteState memory state;
+            state.sender = msg.sender;
+            state.mtx = mtxs[i];
+            state.hash = getMetaTransactionV2Hash(mtxs[i]);
+            state.signature = signatures[i];
+
+            returnResults[i] = _executeMetaTransactionPrivate(state);
+        }
+    }
+
+    /// @dev Get the block at which a meta-transaction has been executed.
+    /// @param mtx The meta-transaction.
+    /// @return blockNumber The block height when the meta-transactioin was executed.
+    function getMetaTransactionV2ExecutedBlock(
+        MetaTransactionDataV2 memory mtx
+    ) public view override returns (uint256 blockNumber) {
+        return getMetaTransactionV2HashExecutedBlock(getMetaTransactionV2Hash(mtx));
+    }
+
+    /// @dev Get the block at which a meta-transaction hash has been executed.
+    /// @param mtxHash The meta-transaction hash.
+    /// @return blockNumber The block height when the meta-transactioin was executed.
+    function getMetaTransactionV2HashExecutedBlock(bytes32 mtxHash) public view override returns (uint256 blockNumber) {
+        return LibMetaTransactionsStorage.getStorage().mtxHashToExecutedBlockNumber[mtxHash];
+    }
+
+    /// @dev Get the EIP712 hash of a meta-transaction.
+    /// @param mtx The meta-transaction.
+    /// @return mtxHash The EIP712 hash of `mtx`.
+    function getMetaTransactionV2Hash(MetaTransactionDataV2 memory mtx) public view override returns (bytes32 mtxHash) {
+        return
+            _getEIP712Hash(
+                keccak256(
+                    abi.encode(
+                        MTX_EIP712_TYPEHASH,
+                        mtx.signer,
+                        mtx.sender,
+                        mtx.expirationTimeSeconds,
+                        mtx.salt,
+                        keccak256(mtx.callData),
+                        mtx.feeToken,
+                        mtx.fees

```

Comment:
This isn't a valid struct hash afaik. Pretty sure the correct way to encode the fees is something like `keccak256(abi.encode(mtx.fees))`, but you should verify. It may be that you need to recursively encode the fee structs with their typehash too. See the [original branch](https://github.com/0xProject/protocol/blob/929d35c30cb20f1f4ce7936539fe7412c8177af1/contracts/zero-ex/contracts/src/features/MetaTransactionsFeature.sol#L225) with this change. Also see [EIP712 encodeData](https://eips.ethereum.org/EIPS/eip-712#definition-of-encodedata).

Reply:
Yeah, that's a good point. The correct way to encode an array of structs is to first encode (hash) each struct individually (with its typehash) and then hash the array. If you're encoding an array of basic types, you just hash the array.

Reply:
This issue should be fixed in the latest version. Jacob and I tested against MetaMask to verify that the signature we are produces matches what MetaMask would produce signing a MetaTransactionDataV2 struct.